packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts

```diff
@@ -0,0 +1,211 @@
+import type { Type } from 'typescript';
+
+import * as util from '../util';
+
+export type Options = [
+  {
+    ignoreIntersections?: boolean;
+    ignoreUnions?: boolean;
+  },
+];
+
+export type MessageIds = 'duplicate';
+
+const astIgnoreKeys = ['range', 'loc', 'parent'];
+
+const isSameAstNode = (actualNode: unknown, expectedNode: unknown): boolean => {
+  if (actualNode === expectedNode) {
+    return true;
+  }
+  if (
+    actualNode &&
+    expectedNode &&
+    typeof actualNode === 'object' &&
+    typeof expectedNode === 'object'
+  ) {
+    if (Array.isArray(actualNode) && Array.isArray(expectedNode)) {
+      if (actualNode.length !== expectedNode.length) {
+        return false;
+      }
+      return !actualNode.some(
+        (nodeEle, index) => !isSameAstNode(nodeEle, expectedNode[index]),
+      );
+    }
+    const actualNodeKeys = Object.keys(actualNode).filter(
+      key => !astIgnoreKeys.includes(key),
+    );
+    const expectedNodeKeys = Object.keys(expectedNode).filter(
+      key => !astIgnoreKeys.includes(key),
+    );
+    if (actualNodeKeys.length !== expectedNodeKeys.length) {
+      return false;
+    }
+    if (
+      actualNodeKeys.some(
+        actualNodeKey =>
+          !Object.prototype.hasOwnProperty.call(expectedNode, actualNodeKey),
+      )
+    ) {
+      return false;
+    }
+    if (
+      actualNodeKeys.some(
+        actualNodeKey =>
+          !isSameAstNode(
+            actualNode[actualNodeKey as keyof typeof actualNode],
+            expectedNode[actualNodeKey as keyof typeof expectedNode],
+          ),
+      )
+    ) {
+      return false;
+    }
+    return true;
+  }
+  return false;
+};
+
+export default util.createRule<Options, MessageIds>({
+  name: 'no-duplicate-type-constituents',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        'Disallow duplicate constituents of union or intersection types',
+      recommended: false,
+      requiresTypeChecking: true,
+    },
+    fixable: 'code',
+    messages: {
+      duplicate:
+        '{{type}} type constituents {{duplicated}} is duplicated with {{previous}}.',
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          ignoreIntersections: {
+            type: 'boolean',
+          },
+          ignoreUnions: {
+            type: 'boolean',
+          },
+        },
+      },
+    ],
+  },
+  defaultOptions: [
+    {
+      ignoreIntersections: false,
+      ignoreUnions: false,
+    },
+  ],
+  create(context, [{ ignoreIntersections, ignoreUnions }]) {
+    const parserServices = util.getParserServices(context);
+    const checker = parserServices.program.getTypeChecker();
+
+    function checkDuplicate(
+      node: TSESTree.TSIntersectionType | TSESTree.TSUnionType,
+    ): void {
+      const cachedTypeMap: Map<TSESTree.TypeNode, Type> = new Map();
+      node.types.reduce<TSESTree.TypeNode[]>(
+        (uniqConstituents, constituentNode) => {
```

Comment:
[Nitpick] I'm not a fan of abbreviating names in code. It makes it harder to read & understand because you have to memorize all sorts of common abbreviations - which can be different across people / codebases.

```suggestion
        (uniqueConstituents, constituentNode) => {
```